### PR TITLE
[To rel/0.13] Improve the process of secondary

### PR DESCRIPTION
**Pull request by ljn55966005**

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/engine/StorageEngine.java [hunk 1]
```diff
@@ -393,6 +404,14 @@ public void start() {
     startTimedService();
   }
 
+  private void checkSecondaryIsLife() {
```
Comment (Beyyes) [inline server/src/main/java/org/apache/iotdb/db/engine/StorageEngine.java:407]:
```suggestion
  private void checkSecondaryAlive() {
```

Comment (Beyyes) [reply server/src/main/java/org/apache/iotdb/db/engine/StorageEngine.java:407]: `life` is a noun

File: server/src/main/java/org/apache/iotdb/db/engine/StorageEngine.java [hunk 2]
```diff
@@ -114,6 +97,7 @@ public class StorageEngine implements IService {
 
   private static final IoTDBConfig config = IoTDBDescriptor.getInstance().getConfig();
   private static final long TTL_CHECK_INTERVAL = 60 * 1000L;
+  private static final long HEARTBEAT_CHECK_INTERVAL = 30L;
```
Comment (Beyyes) [inline server/src/main/java/org/apache/iotdb/db/engine/StorageEngine.java:100]: If it's too long to check the alive info of Secondary in every 30 seconds

File: server/src/main/java/org/apache/iotdb/db/doublelive/OperationSyncProducer.java [hunk 3]
```diff
@@ -33,22 +35,47 @@
...
+      Pair<ByteBuffer, OperationSyncPlanTypeUtils.OperationSyncPlanType> planPair,
+      String deviceName) {
+    ByteBuffer headBuffer;
+    headBuffer = planPair.left;
+    headBuffer.position(0);
+    int index;
+    if (deviceName == null) {
+      index = operationSyncQueues.size();
+    } else {
+      index = deviceName.hashCode() % (operationSyncQueues.size() - 1);
```
Comment (Beyyes) [inline server/src/main/java/org/apache/iotdb/db/doublelive/OperationSyncProducer.java:62]: The result of deviceName.hashCode() may be negative

File: server/src/main/java/org/apache/iotdb/db/doublelive/OperationSyncConsumer.java [hunk 4]
```diff
@@ -39,10 +40,10 @@ public class OperationSyncConsumer implements Runnable {
 
   public OperationSyncConsumer(
       BlockingQueue<Pair<ByteBuffer, OperationSyncPlanTypeUtils.OperationSyncPlanType>>
-          OperationSyncQueue,
+          operationSyncQueue,
       SessionPool operationSyncSessionPool,
       OperationSyncLogService dmlLogService) {
-    this.OperationSyncQueue = OperationSyncQueue;
+    this.OperationSyncQueue = operationSyncQueue;
```
Comment (Beyyes) [inline server/src/main/java/org/apache/iotdb/db/doublelive/OperationSyncConsumer.java:46]: this.operationSyncQueue
The first letter should be lower case.

File: server/src/main/java/org/apache/iotdb/db/doublelive/OperationSyncProducer.java [hunk 5]
```diff
@@ -33,22 +35,52 @@
...
+  public void put(
+      Pair<ByteBuffer, OperationSyncPlanTypeUtils.OperationSyncPlanType> planPair,
+      String deviceName) {
+
+    ByteBuffer headBuffer;
+    headBuffer = planPair.left;
+    headBuffer.position(0);
+    int index;
+    if (deviceName == null) {
+      index = operationSyncQueues.size() - 1;
```
Comment (CRZbulabula) [inline server/src/main/java/org/apache/iotdb/db/doublelive/OperationSyncProducer.java:61]: Would it be better if we use random selection here?

File: server/src/main/java/org/apache/iotdb/db/doublelive/OperationSyncProducer.java [hunk 6]
```diff
@@ -33,22 +35,52 @@
...
+    } else {
+      try {
+        index = Math.abs(deviceName.hashCode()) % (operationSyncQueues.size() - 1);
+      } catch (Exception e) {
+        index = 1;
+      }
```
Comment (CRZbulabula) [inline server/src/main/java/org/apache/iotdb/db/doublelive/OperationSyncProducer.java:67]: I wonder in which case will this block of code throw an Exception? And the same as before, would it be better if we use random selection when some Exceptions are thrown?

File: server/src/main/java/org/apache/iotdb/db/doublelive/OperationSyncProducer.java [hunk 7]
```diff
@@ -33,22 +35,52 @@
...
+    }
+    for (int i = 0; i < RETRY; i++) {
+      // retry 3 times
+      if (operationSyncQueues.get(index).offer(planPair)) {
+        return;
+      }
+    }
```
Comment (CRZbulabula) [inline server/src/main/java/org/apache/iotdb/db/doublelive/OperationSyncProducer.java:74]: It's better to sleep 1s or less between each loop of retry.

File: server/src/main/java/org/apache/iotdb/db/engine/StorageEngine.java [hunk 8]
```diff
@@ -87,33 +86,17 @@
...
-import java.util.Set;
-import java.util.concurrent.ArrayBlockingQueue;
-import java.util.concurrent.BlockingQueue;
-import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.ExecutionException;
-import java.util.concurrent.ExecutorService;
-import java.util.concurrent.Future;
-import java.util.concurrent.ScheduledExecutorService;
-import java.util.concurrent.TimeUnit;
+import java.util.concurrent.*;
```
Comment (CRZbulabula) [inline server/src/main/java/org/apache/iotdb/db/engine/StorageEngine.java:91]: We should always avoid to `import *`

File: server/src/main/java/org/apache/iotdb/db/engine/StorageEngine.java [hunk 9]
```diff
@@ -122,6 +105,7 @@ public class StorageEngine implements IService {
   private static OperationSyncProducer operationSyncProducer;
   private static OperationSyncDDLProtector operationSyncDDLProtector;
   private static OperationSyncLogService operationSyncDDLLogService;
+  private static boolean isSecondaryLife;
```
Comment (CRZbulabula) [inline server/src/main/java/org/apache/iotdb/db/engine/StorageEngine.java:108]: This field should better to be set to AtomicBoolean or volatile for thread-safety

**Review by CRZbulabula — CHANGES_REQUESTED** [review]
PTAL

File: server/src/main/java/org/apache/iotdb/db/doublelive/OperationSyncProducer.java (hunk 5 above)
Comment (ljn55966005) [reply server/src/main/java/org/apache/iotdb/db/doublelive/OperationSyncProducer.java:61]: we must prove the order of the same device

File: server/src/main/java/org/apache/iotdb/db/doublelive/OperationSyncProducer.java (hunk 6 above)
Comment (ljn55966005) [reply server/src/main/java/org/apache/iotdb/db/doublelive/OperationSyncProducer.java:67]: when the queue size is 1

**Review by Beyyes — APPROVED** [review]
LGTM

**Review by qiaojialin — APPROVED** [review]

**Pull request merged**